Add lasso tool for faster drawing of polygonal Shapes

LucaMarconato opened the pull request:
# Description

This PR adds a polygon lasso tool for drawing complex polygons which are added to a shape layer. The tool does not require the user to click for adding each specific vertex as is the case with the polygon tool. This PR features the following:
• A new shape layer button which can be activated by keyboard shortcut shift + P
• A polygon lasso mode which can be used for drawing with the mouse and tablet

For mouse draw mode the sequence of events by the user is as follows:

1. Mouse press
2. Move mouse while not mouse pressed
3. Mouse press to finish draw

For tablet mode, the shape is created by a mouse (ie stylus) click+drag event. Drawing finishes when the stylus is lifted / on mouse release.

For both modes, vertices are added only if the vertex to be added is at least 10 screen pixels away from the previous vertex. This is a setting configurable in advanced settings.

An implementation of the Ramer–Douglas–Peucker algorithm is used to reduce the number of vertices when a drawing is finished. The aggressiveness of reducing the number of vertices is dependent on a parameter epsilon. If set to 0, no vertices are removed. The higher the epsilon, the more vertices are removed. The epsilon parameter can be configured in napari preferences -> experimental -> rdp_epsilon.

Lastly, a refactor has been performed for various functions related to polygon drawing.

# References

We acknowledge @yozhikoff who made a first prototype of this new feature.

LucaMarconato committed:
adding lasso tool for drawing polygonal shapes

LucaMarconato committed:
wip support for mouse draw with mouse press

LucaMarconato committed:
repositioned button

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -222,6 +232,27 @@ def add_path_polygon_creating(layer, event):
+        _move(layer, coordinates)
+
+        now = datetime.datetime.now()
+        MAX_POINTS_PER_SECOND = 60
```
My instinct tells me that this should be distance based rather than time-based. Specifically, canvas distance (say, 5 pixels), to make sure zooming in and out stays intuitive.

melonora replied:
I agree

brisvag reviewed (commented):
Love this! I was actually trying to do the same a few weeks ago, but got stuck, so it's great to see this :)

In general, I would prever to have this as an alternative behaviour of the polygon tool. I love option `1.`, it feels very intuitive. It's indeed a bit harder to implement, but hopefully you can get past where I got stuck :P

One thing to note is that I recently added [spline interpolation](https://github.com/napari/napari/pull/5334) to the polygon shape, which is currently not exposed by should work if exposed. There is some discussion in that PR (and its predecessors) about how to expose this exactly, but nothing conclusive. It might be useful to consider here, since I expect most use cases where lasso drawing is useful would prefer to have smooth shapes rather than jagged. (note however that the smoothing is only visual for now, and there is no way to currently get it out at higher levels, for example to generate labels).

melonora committed:
Use cursor position delta

melonora committed:
Add RDP polygon point reduction

melonora committed:
distance function works to any dimension

melonora committed:
Merge pull request #1 from melonora/polygon_lasso

Add RDP point reduction

melonora commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -198,10 +207,19 @@ def add_path_polygon(layer, event):
+    elif event.type == 'mouse_press' and layer._mode == Mode.ADD_POLYGON_LASSO:
+        index = layer._moving_value[0]
+        vertices = layer._data_view.shapes[index].data
+        vertices = rdp(vertices, epsilon=0.5)
```
One point of discussion is whether we should expose this epsilon parameter to the user somehow.

LucaMarconato replied:
Not sure I would expose, but I would make some empirical tests of drawing curves slower or faster and with more or less curvature/sharp corners/wiggling, and empirically find a reasonable epsilon.

melonora replied:
I did that for regular cell shapes. However, the choice of this parameter does depend on what you are trying to annotate and how accurate you need it to be.

Drawing slower or faster does not matter at all. The vertices are added based on a distance threshold of 5 pixels.

LucaMarconato commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -1144,3 +1149,70 @@ def validate_num_vertices(
             )
+
+
+def perpendicular_distance(
```
lgtm

LucaMarconato commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -1144,3 +1149,70 @@ def validate_num_vertices(
+        return np.linalg.norm(t * (line_start - line_end) + line_end - point)
+
+
+def rdp(vertices: npt.NDArray, epsilon: float) -> npt.NDArray:
```
I don't know the algorithm, so I skip the review of this function.

brisvag reviewed (commented):
Am I understanding correctly that the `rdp` use here is changing the actual points used to define the shape? In that case, I don't think it's mutually exclusive with the spline interpolation (which is only a visualisation tool).

melonora committed:
Fixed broadcasting error

Error occured when drawing in different dimension beforehand due to last_cursor_position not being reset

melonora committed:
Combined new and old polygon draw functionality

- Combined the old way of creating a polygon with the new lasso drawing
- Allow for multiple callbacks per mode enum with a given layer mode
- Refactor

melonora committed:
Fix add path not working

melonora committed:
Adjust threshold

melonora committed:
Merge pull request #2 from melonora/polygon_lasso

- Combined the old polygon draw functionality with the new lasso draw functionality. Both functionalities are combined in the same mode (add polygon)
- Allow for multiple callbacks for a given layer mode and mouse mode.
- Refactor
-  Added docstrings

melonora committed:
Fix rdp not adjusting data

melonora committed:
Merge pull request #3 from melonora/polygon_lasso

Fix rdp not adjusting data

melonora committed:
Removed polygon lasso button functionality

melonora committed:
Ensure no points removed with epsilon equal to 0

melonora committed:
Brang back polygon_lasso button

melonora committed:
Bring back polygon_lasso button

melonora committed:
Merge branch 'polygon_lasso' of https://github.com/melonora/napari into polygon_lasso

melonora committed:
Revert "Bring back polygon_lasso button"

This reverts commit c90e0462f6377e4cb8e3a5e5a637e667ac5fcfc9.

melonora committed:
Revert "Removed polygon lasso button functionality"

This reverts commit 9f6d19c6ea604c2b899496c2d15744427f18db3b.

melonora committed:
Revert "Fix add path not working"

This reverts commit 06b51f643844f7f4bca5293eb7b95c1a984a0a2b.

melonora committed:
Revert to fixed rdp with original lasso button

melonora committed:
Merge main

melonora committed:
Add lasso tablet button

melonora committed:
New button with duplicate lasso functionality

melonora committed:
Add tablet draw functionality

melonora committed:
Clarify distance threshold

melonora committed:
Merge pull request #4 from melonora/polygon_lasso

Revert to 3 buttons, one for original add polygon functionality, one for mouse draw polygon functionality and one for tablet draw polygon functionality. Distance threshold is now also calculated based on screen position instead of canvas position.

melonora committed:
Let tablet mode respect distance threshold

melonora committed:
Merge pull request #5 from melonora/polygon_lasso

Let tablet mode respect distance threshold

melonora committed:
Remove TODO

melonora committed:
Prevent meshes from being drawn while creating lasso polygon

melonora committed:
Fix color of pen in icon

melonora committed:
Merge pull request #6 from melonora/polygon_lasso

Prevent meshes from being drawn while drawing polygons in both tablet and mouse mode. Also fix color of pen in icon.

melonora committed:
Merge branch 'main' into polygon_lasso

melonora committed:
Add shortcut for polygon_lasso

melonora committed:
no-op for double click mode polygon_lasso

melonora committed:
Combine mouse and tablet draw in one button

melonora committed:
Merge pull request #7 from melonora/polygon_lasso

Both drawing modes are now combined in one button. At the moment this works without problems except that the distance threshold is not respected.

melonora committed:
clean up commented code

melonora committed:
Merge pull request #8 from melonora/polygon_lasso

Clean up commented code and fixed button order

melonora committed:
Fix respecting distance threshold

- Tiny refactor

melonora committed:
Merge pull request #9 from melonora/polygon_lasso

Fix respecting distance threshold
Tiny refactor

melonora committed:
Initial refactor

melonora committed:
remove event parameter

melonora committed:
Bugfix: finish shape with 2 vertices

melonora committed:
Refactor

- put redundant code in functions
- added docstrings

melonora committed:
Merge pull request #10 from melonora/polygon_lasso

This is a refactor which eliminates the use of redundant code and puts duplicated code in separate functions and adds docstrings and type hints.

melonora committed:
Merge branch 'main' into polygon_lasso

melonora committed:
Allow for changing epsilon parameter in experimental settings

melonora committed:
Merge pull request #11 from melonora/polygon_lasso

Added RDP epsilon to application experimental settings.

melonora committed:
Add rdp test

melonora committed:
Add test for perpendicular distance

melonora committed:
move line start, end, points  out of parametrize

melonora committed:
Add polygon_lasso tests

melonora committed:
Merge pull request #12 from melonora/polygon_lasso

Added tests and adjusted events in shapes_mouse_bindings tests in order for event.pos to be supported

melonora committed:
Disallow epsilon lower than 0

melonora committed:
Merge pull request #13 from melonora/polygon_lasso

Disallow epsilon lower than 0

melonora committed:
Merge branch 'main' into polygon_lasso

psobolewskiPhD commented on napari/settings/_experimental.py:
```diff
@@ -37,6 +37,16 @@ class ExperimentalSettings(EventedSettings):
 
+    rdp_epsilon: float = Field(
+        0.5,
+        title=trans._("RDP epsilon"),
```
```suggestion
        title=trans._("Shapes polygon lasso RDP epsilon"),
```

melonora committed:
Update napari/settings/_experimental.py

make title more informative

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

melonora committed:
Allow stop draw by escape press

melonora committed:
Merge pull request #14 from melonora/polygon_lasso

Allow for stop drawing the polygon shape by pressing escape in correspondence with napari docs.

melonora committed:
Fix test by changing subsetting and not running rdp twice

melonora committed:
Merge pull request #15 from melonora/polygon_lasso

Fix test by changing subsetting and not running rdp twice

melonora committed:
remove statement

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -100,8 +136,20 @@ def select(layer, event):
-def add_line(layer, event):
-    """Add a line."""
+def add_line(layer: Shapes, event: ReadOnlyWrapper) -> None:
+    """Add a line.
```
I know this isn't your function, but, since you've annotated it I presume (?) that you have a bit more of a grasp about what it does? If so, could you please explain in an extra paragraph *how* it adds a line? ie from where to where and (if it makes sense) how wide.

melonora replied:
The edge width is purely a visual which is documented elsewhere. I added to the docstring, it basically starts with 2 points in data space where the second point has a minor offset to the first one if mouse has not moved yet. When the mouse starts moving the second point will track the cursor. Not certain whether this offset is required as someone could misclick and then just release. Now this would create a tiny line while I personally would expect that this would delete the line just as when you have a polygon with less than 3 vertices. This would be something to for an issue / separate PR though.

brisvag replied:
AFAIK The offset is requires because otherwise you'd get a degenerate 0-length line when you start drawing (since the end of line is constantly updated to the last mouse position) and this can mess up the math later.

melonora replied:
How would a 0-length line be different from the situation of starting to draw a polygon? A user clicks and 2 points are being created, 1 starting point and one active point which before moving the mouse is at exactly the same coordinate.
 
If these have the same coordinates there should be no problem while `layer._is_creating` is `True`. If finishing drawing a line like this should just be deleted as the user probably initiated the draw by mistake.

jni replied:
@melonora could you make a new issue? In the meantime this can be resolved, although I'd prefer if the docstring was wrapped to 80c like the rest of the code. 😅

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -12,18 +15,51 @@
+    import numpy.typing as npt
+
+    from napari.layers.shapes.shapes import Shapes
+    from napari.utils._proxies import ReadOnlyWrapper
```
I find this type annotation misleading to be honest, and worse than nothing. We *incidentally* happen to wrap events before sending them to callbacks, but the important part of the type is that it's a MouseEvent, not a ReadOnlyWrapper. We might as well write `object`. My suggestion would be to do one of the following:

1. Simply remove the annotation; in the docstring, don't write `event: ReadOnlyWrapper` but rather, `event: ReadOnlyWrapper(vispy.MouseEvent)`. (Or whatever the actual type is.)
2. Find a way to make ReadOnlyWrapper a generic type, so that you can annotate as `ReadOnlyWrapper[vispy.MouseEvent]`.
3. Define a MouseEventProto [PEP-544 protocol](https://peps.python.org/pep-0544/#using-protocols), which both the MouseEvent and the wrapped MouseEvent comply with, and annotate with that. However I don't know whether MyPy is smart enough to find that the wrapped object indeed satisfies the protocol.

My preference would be for 1. As I'm not super familiar with typing I don't know how to gauge the feasibility of 2 or 3. Opinions from @Czaki and @tlambert03 here would be useful! 🙏

tlambert03 replied:
Yeah, 2 seems like the correct approach, but 1 seems like the "easy" and just fine approach :)

jni replied:
@tlambert03 is 2 "hard"? I have no experience here. (And potentially the "correct" way is to upstream to wrapt, right?)

melonora replied:
One thing to note is that from what I heard from @Czaki these events should not be wrapped. I will correct the annotation for now though given that this is just the current situation.

tlambert03 replied:
it's not hard to make it a Generic (you can look at the `PublicOnlyProxy` that I did in the same module, which is already a `Generic`) ... but it _is_ hard to make it all that useful for the sake of typing.  For example, if you do this, you still don't benefit much from the typing perspective:

```python
def select(layer: Shapes, event: ReadOnlyWrapper[MouseEvent]) -> None:
    event.modifiers  # mypy still says "Any"...
```

you kinda need to lie even more to the typing system to make it realize that any `__getattribute__` call will return the type of the `__wrapped__` object.  See https://github.com/python/typing/issues/802 for more discussion.

If I were you, I'd just type this as the wrapped object directly:

```py
from vispy.app.canvas import MouseEvent

def select(layer: Shapes, event: MouseEvent) -> None:
    ...
```

from the end-user perspective, that's totally correct.  From an internal/napari perspective, you might need to use some type ignores for whatever internally *calls* the `select` function with a proxy object ... but that's no biggie

melonora replied:
I would say lets go for the direct type annotation as MouseEvent as this should not be an ReadOnlyWrapper in the first place and thus would be fixed in the future.

melonora replied:
Resolved with [48b7970](https://github.com/napari/napari/pull/5555/commits/48b7970bf8033fa945a41a69b140f45f8a897900). Though ReadOnlyWrapper is not the type hint I still do mention it in the docstring for clarity that this is not a pure MouseEvent and to remind that this should be changed.

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -12,18 +15,51 @@
-    """Highlight hovered shapes."""
+def highlight(layer: Shapes, event: ReadOnlyWrapper) -> None:
+    """Highlight hovered shapes, including boundaries, vertices, interaction boxes, and drag
+        selection box when appropriate.
```
PSA: there is a PEP about docstring formatting, [PEP-257](https://peps.python.org/pep-0257/), which [says](https://peps.python.org/pep-0257/#multi-line-docstrings) (emphasis mine):

> One-liners are for really obvious cases. They should really fit on one line. *[80c]*
> [...]
> The docstring is a phrase ending in a period. It prescribes the function or method’s effect as a command (“Do this”, “Return that”), not as a description; e.g. don’t write “Returns the pathname …”.
> [...]
> **Multi-line docstrings consist of a summary line just like a one-line docstring, followed by a blank line, followed by a more elaborate description.** The summary line may be used by automatic indexing tools; it is important that it **fits on one line and is separated from the rest of the docstring by a blank line.**

fwiw I'd be happy if we could add this requirement to our linters @Czaki!

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -12,18 +15,51 @@
+    event: ReadOnlyWrapper
+        A proxy read only wrapper around a mouse event. Though not used here it is passed as argument by the shapes
+        layer mouse move callbacks.
+
+    Returns
+    -------
+    None
```
I think None returns are optional to list; personally I prefer leaving them out, but either way it should be consistent (add_ellipse below doesn't have one).

melonora replied:
We can leave them out. It isn't mentioned in PEP or numpy docstring, though numpy does leave them out of the docstring as well.

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -117,8 +165,16 @@ def add_line(layer, event):
+    layer: Shapes
+        Napari shapes layer
+    event: ReadOnlyWrapper
+        A proxy read only wrapper around a mouse event"""
```
```suggestion
        A proxy read only wrapper around a mouse event.
    """
```

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -117,8 +165,16 @@ def add_line(layer, event):
-def add_ellipse(layer, event):
-    """Add an ellipse."""
+def add_ellipse(layer: Shapes, event: ReadOnlyWrapper):
+    """
+    Add an ellipse to the shapes layer.
```
```suggestion
    """Add an ellipse to the shapes layer.
```

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+def finish_drawing_shape(layer: Shapes, event: ReadOnlyWrapper) -> None:
     """
-    finish drawing the current shape
+    Calls the finish drawing method of the shapes layer which resets all the properties used for shape drawing
+    and deletes the shape if the number of vertices do not meet the threshold of 3.
```
See my PEP-257 PSA above

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+    layer: Shapes, coordinates: Tuple[np.float, ...]
+) -> None:
+    """Creating the shape when initializing the draw, adding and selecting the initiatlized shape and
+    setting required layer attributes for drawing.
```
See my PEP-257 PSA above

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+
+def add_path_polygon_lasso(layer: Shapes, event: ReadOnlyWrapper) -> None:
+    """Function responsible for initiating, drawing and finishing the lasso polygon in drag mode (tablet) or
+    initiating and finishing the lasso polygon when drawing with the mouse.
```
See my PEP-257 PSA above

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+    coordinates: Tuple[np.float, ...],
+    new_type: Optional[str],
+) -> None:
+    """Adds a vertex to an existing path or polygon and edits the layer view.
```
```suggestion
    """Add a vertex to an existing path or polygon and edits the layer view.
```

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+            position_diff = np.linalg.norm(
+                event.pos - layer._last_cursor_position
+            )
+            if position_diff > 10:
```
@melonora should this value (10) be a preference?

Czaki replied:
Agree with jni

jni replied:
@melonora do you have time to add a preference here? 🙏 Having said that @melonora @Czaki I think it would be ok to merge without that and make an issue about it / fix in followup PR. This is already a great PR adding great functionality.

melonora replied:
I can add it to the settings, wouldn't be a problem. Sorry I missed this comment initially.

melonora replied:
Resolved with [0d2e01c](https://github.com/napari/napari/pull/5555/commits/0d2e01c69ab129fc9e399f502e80bacd72412ff6). I will also update the docs PR and PR description.

melonora replied:
Ah nvm regarding the PR description. I can't update it myself and @LucaMarconato is on holiday at the moment. If you, @Czaki or @jni, could do this, it would be much appreciated. Basically add a line:

Exposing the vertex distance threshold in napari preferences -> experimental -> lasso_vertex_distance.
The wheel cannot be set lower than 1 and not higher than 50. Default value is 10. 

Along with some changes in the text above to already indicate that 10 screen pixels is a setting.

melonora replied:
@psobolewskiPhD I updated the docs PR, could you have a look please?

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+
+def polygon_creating(layer: Shapes, event: ReadOnlyWrapper) -> None:
+    """Function providing the functionality of moving the vertex when moving the mouse and adding it if the distance
+    threshold criterium is met.
```
It took me a while to grok this, because the vertex is not the last-placed vertex, but rather the "leading" or "virtual" vertex at the drawing front. Here's my suggestion:

```suggestion
    """Follow mouse while drawing, placing a polygon vertex after a certain distance.

    When drawing a polygon in lasso mode, a vertex follows the cursor, creating a polygon
    visually that is *not* the final polygon to be created: it is the polygon if the current
    mouse position were to be the last position added. After the mouse moves a certain
    distance (configurable in Settings), a new vertex is automatically added and the
    distance counter resets.
```

Does that sound about right? I would also specify the complete setting path. Also, I wrote this on the GitHub UI so I don't know if my lines are too long... They probably are so it'll need some tweaking.

melonora replied:
Currently this distance is not exposed in the settings. The RDP epsilon is a different distance cut off. The distance for adding vertices could be exposed though. One thing though is that this distance is dependent on the number of pixels on the screen. Hence when zooming in, a lot more datapoints will be added for a smaller unit in data space.

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+
+
+def add_path_polygon(layer: Shapes, event: ReadOnlyWrapper) -> None:
+    """Add a path or polygon or add vertex to an existing one.
```
Can you elaborate on how this works in a paragraph? It's unclear looking at the signature how it should work, which makes reading the function harder.

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -178,53 +263,216 @@ def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+    layer: Shapes
+        Napari shapes layer
+    coordinates: Tuple[np.float, ...]
+        The coordinates of the vertex to be potentially added, e.g. vertex tracks the mouse cursor position.
```
Specify the coordinate space of these coordinates? It's data coordinates, right?

Anyway, it's a bit confusing that the base function is called `_move` and takes in a layer and a set of coordinates... Could we rename it to e.g. `_move_last_vertex` or something like that?

jni replied:
Oh, just saw the function, it's more general than that... 😅 Um, `_move_active_element_under_cursor`?

melonora replied:
Will be resolved with a PR to Luca's branch this evening. Changed to _move_active_element_under_cursor.

jni commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -358,23 +623,44 @@ def _drag_selection_box(layer, coordinates):
+    layer: Shapes, coordinates: Tuple[np.float, ...]
+) -> List[np.float, ...]:
+    """Sets the coordinates relative to the center of the bounding box of a shape and returns the position
+    of where a drag event of a shape started.
```
PEP257

jni commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -1140,3 +1145,74 @@ def validate_num_vertices(
+    point: npt.NDArray, line_start: npt.NDArray, line_end: npt.NDArray
+) -> npt.float:
+    """Function to calculate the perpendicular distance of a point to a given euclidean line defined
+    by 2 points, line_start and line_end. Works up to any dimension
```
PEP257

jni reviewed (changes requested):
@melonora @LucaMarconato wow this is very close! 🙏 I pulled down the changes to try things out and the feel is 😍 just wonderful!

I've left some rather nitpicky changes about docstring formatting and type annotation, but basically this is ready minus those nits imho! I'm really impressed by how much simpler the implementation is compared to a week ago! 😂 👏

melonora committed:
minor docstring refactor

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

melonora committed:
Merge branch 'napari:main' into polygon_lasso

melonora committed:
Merge branch 'napari:main' into polygon_lasso

melonora committed:
Refactor function name

melonora committed:
Refactor add_line docstring

melonora committed:
Refactor ReadOnlyWrapper typehints

melonora committed:
Refactor None return docstring

melonora committed:
Refactor docstring PEP257

melonora committed:
Refactor docstring in correspondence with PEP257

melonora committed:
Merge branch 'polygon_lasso' into polygon_lasso

melonora committed:
Merge pull request #16 from melonora/polygon_lasso

Refactor docstrings

melonora committed:
Function name refactor

melonora committed:
Merge branch 'polygon_lasso' of https://github.com/melonora/napari into polygon_lasso

melonora committed:
Merge pull request #17 from melonora/polygon_lasso

Function name refactor

melonora committed:
Merge branch 'main' into polygon_lasso

Czaki commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -334,15 +568,17 @@ def vertex_remove(layer, event):
 
-def _drag_selection_box(layer, coordinates):
+def _drag_selection_box(
+    layer: Shapes, coordinates: Tuple[np.float, ...]
```
It is required to have `np.float` not normal `float`?

jni replied:
And incidentally, `np.float` *is* `float` and is deprecated:

```python
In [1]: np.float is float
<ipython-input-1-09ec2f1d7cc2>:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.float is float
Out[1]: True
```

melonora replied:
resolved with [3fdedf8](https://github.com/napari/napari/pull/5555/commits/3fdedf82c7fe2b36889ad347842451f40178db40) and [3b987c1](https://github.com/napari/napari/pull/5555/commits/3b987c12fd70e0f60b57363ecbb697351411405e)

Czaki commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -358,22 +594,43 @@ def _drag_selection_box(layer, coordinates):
 
-def _set_drag_start(layer, coordinates):
+def _set_drag_start(
+    layer: Shapes, coordinates: Tuple[np.float, ...]
```
same as above

melonora replied:
Could be that it would, I type hinted here with what is specifically event.position

melonora replied:
resolved with [3fdedf8](https://github.com/napari/napari/pull/5555/commits/3fdedf82c7fe2b36889ad347842451f40178db40) and [3b987c1](https://github.com/napari/napari/pull/5555/commits/3b987c12fd70e0f60b57363ecbb697351411405e)

Czaki commented on napari/layers/shapes/_shapes_mouse_bindings.py:
```diff
@@ -501,16 +758,25 @@ def _move(layer, coordinates):
+        if shape_type == Ellipse:
+            # DIRECT vertex moving of ellipse not implemented
+            pass
+        else:
```
Maybe this change should be reverted? RAter plain TODO should be added.

melonora replied:
I agree. Not certain though why this shows up as being changed though. The same lines exist on main: [code section](https://github.com/napari/napari/blob/d8dd1d2f592618064d2b5ac0c770ac1eb4741e77/napari/layers/shapes/_shapes_mouse_bindings.py#LL234C1-L237C1).

melonora replied:
resolved with [7b0cb08](https://github.com/napari/napari/pull/5555/commits/7b0cb08405cec16ed2f4f8141ca9712869cc022f)

Czaki reviewed (commented):
Few remarks. I will try to make a more detailed review tomorrow/Friday.

melonora committed:
np.float -> float

melonora committed:
np.float  -> float

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

melonora committed:
Convert comment to TODO

melonora committed:
Merge branch 'main' into polygon_lasso

melonora committed:
Merge branch 'main' into polygon_lasso

melonora committed:
Merge branch 'main' into polygon_lasso

melonora committed:
Merge branch 'main' into polygon_lasso

melonora committed:
Merge branch 'napari:main' into polygon_lasso

melonora committed:
Add distance threshold to settings.

melonora committed:
Merge pull request #18 from melonora/polygon_lasso

Add vertex distance threshold to settings.
